File: functions/go-functions/reducer/reducer/handler.go
package function

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"time"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

type CONT struct {
	BucketName   string `json:"bucketName"`
	Key          string `json:"key"`
	OutputBucket string `json:"outputBucket"`
}

type ValueData struct {
	Value int `json:"value"`
}

// Handle a serverless request
func Handle(req []byte) string {

	var cc CONT
	var err error
	err = json.Unmarshal(req, &cc)
	if err != nil {
		fmt.Println(err)
	}

	endpoint := "192.168.178.250:9000"
	accessKeyID := "minioadmin"
	secretAccessKey := "minioadmin"
	useSSL := false

	// Initialize minio client object.
	minioClient, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		log.Fatalln(err)
	}

	var amount int
	startTime := time.Now()

	// Listing objects
	objectCh := minioClient.ListObjects(context.TODO(), cc.BucketName, minio.ListObjectsOptions{
		Prefix:    "key/" + cc.Key,
		Recursive: true,
	})

	// Iterating over the listed objects
	for object := range objectCh {
		if object.Err != nil {
			log.Fatalln(object.Err)
		}

		// Getting object data
		object, err := minioClient.GetObject(context.TODO(), cc.BucketName, object.Key, minio.GetObjectOptions{})
		if err != nil {
			log.Fatalln(err)
		}
		data, err := ioutil.ReadAll(object)
		if err != nil {
			log.Fatalln(err)
		}

		var line ValueData
		err = json.Unmarshal(data, &line)
		if err != nil {
			log.Fatalln(err)
		}

		amount += line.Value
	}

	// Creating output object
	outputData, err := json.Marshal(map[string]int{cc.Key: amount})
	if err != nil {
		log.Fatalln(err)
	}
	_, err = minioClient.PutObject(context.TODO(), cc.OutputBucket, cc.Key, bytes.NewReader(outputData), int64(len(outputData)), minio.PutObjectOptions{
		ContentType: "application/json",
	})
	if err != nil {
		log.Fatalln(err)
	}

	endTime := time.Now()

	// Preparing the response
	response := map[string]interface{}{
		"key":  cc.Key,
		"time": endTime.Sub(startTime).Seconds(),
	}
	responseData, err := json.Marshal(response)
	if err != nil {
		log.Fatalln(err)
	}

	return fmt.Sprintf(string(responseData))
}

File: functions/reducer/customer-reducer/handler.py
def handle(req):

    from minio import Minio
    import json
    import io
    import random
    import string
    import time

    context = json.loads(req)

    client = Minio(
        endpoint="192.168.178.250:9000",
        access_key="minioadmin",
        secret_key="minioadmin",
        secure=False
    )

    my_name = context["bucketName"]
    my_key = context["key"]
    outputbucket = context["outputBucket"]
    #my_name = "test-bucket"
    #my_key = "2"
    #outputbucket = "test-bucket-output"
    data = []
    amount = 0
    start_time = time.time()


    input_data_list = client.list_objects(my_name, prefix="key/"+my_key, recursive=True)
    for obj in input_data_list:
        response = client.get_object(my_name, obj.object_name)
        line = json.loads(response.data.decode())
        amount = amount + int(line["value"])    
                
    client.put_object(outputbucket, my_key, io.BytesIO(json.dumps({my_key : amount}).encode('utf-8')), length=-1, part_size=10*1024*1024 , content_type= "application/json")

    end_time = time.time()
    return json.dumps({
        "key" : my_key,
        "time" : end_time - start_time
    })


def emitValues(outputDict, client):
    return True
